### DISCUSSION TITLE
Solution with IPromotableSinglePhaseNotification and TransactionCompletion against a promoted transaction to MS-DTC

### DISCUSSION THREAD

**Discussion by DavoudEshtehari** [dotnet/runtime#96067] [Q&A]
Based on the transaction documents, subscribing to the [TransactionCompleted](https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Flearn.microsoft.com%2Fen-us%2Fdotnet%2Fapi%2Fsystem.transactions.transaction.transactioncompleted%3Fview%3Dnet-7.0&data=05%7C02%7Cv-deshtehari%40microsoft.com%7Cfe103ff65eba4a8ee9cf08dbfd19e958%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C638382060041617078%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=82kaSZciQd7%2FxVBZDFL3Yp3%2F8HneEi7NycdO6rPQ2Uo%3D&reserved=0) event has negative affection on perf. I'd appreciate any thoughts with the following questions:

1. Is there any opportunity to improve it or this is the best version of the event driven design?
2. What would be the replacement to receive the transaction completion signal with a promoted transaction to DTC through a [IPromotableSinglePhaseNotification](https://learn.microsoft.com/en-us/dotnet/api/system.transactions.ipromotablesinglephasenotification?view=net-7.0) implementation?

Here is the links to the design on MDS: 

- [SqlDelegatedTransaction](https://github.com/dotnet/SqlClient/blob/main/src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/SqlDelegatedTransaction.cs)
- [TransactionCompletedEvent](https://github.com/dotnet/SqlClient/blob/04b531bab5ad5c0bdf204d443b6f3da88c2429b9/src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/ProviderBase/DbConnectionInternal.cs#L472)

**Comment by roji** [comment]
@DavoudEshtehari missed this - discussions (as opposed to issues) don't automatically cause runtime area owner to be pinged, so if you create one, it's a good idea to ping the relevant people manually. In any case, this was cross-posted as #96881 - I'll answer there.

**Reply by DavoudEshtehari** [reply]
@roji I always appreciate your valuable feedback. After reviewing your comments, I was inspired to explore an event-free solution. Currently, I'm exploring the use of promoted connections to DTC. Enlisting in a transaction can be achieved by implementing the `ISinglePhaseNotification` interface through trx.[EnlistVolatile](https://learn.microsoft.com/en-us/dotnet/api/system.transactions.transaction.enlistvolatile), effectively bypassing the transaction completion event, as recommended in the documentation. However, in dealing with a promoted transaction, I'm facing the challenge of needing to implement the `IPromotableSinglePhaseNotification` interface to make a request through trx.[EnlistPromotableSinglePhase](https://learn.microsoft.com/en-us/dotnet/api/system.transactions.transaction.enlistpromotablesinglephase?view=net-8.0), without the option of subscribing to the trx.`TransactionCompleted` event. I'm still eager on finding a resolution in this context to sidestep the involvement of events in transactions.

**Reply by roji** [reply]
I'm happy to hear you're looking at removing the event; I'm guessing there's a reason why subscription to it was added in the first place - maybe it's a good idea to use git blame, and explore the git history where this was added to try to understand it, in order to have the full context?

In any case, note that SqlClient should be enlisting via EnlistDurable and not EnlistVolatile - the latter is meant for cases where recovery isn't supported, because the database/resource has no durable storage etc (e.g. in-memory caches) - this isn't the SQL Server case.

In any case, I'm not sure I understand the problem you're running into... When calling EnlistPromotableSinglePhase(), you pass in a something which implements the IPromotableSinglePhaseNotification interface (in the SqlClient, IIRC that's SqlDelegatedTransaction). That has callbacks for transaction state changes (commit, rollback, promote), which you can react to. All that is already done by SqlClient (otherwise it wouldn't have supported distributed transactions in the first place), including the promotion logic and the state changes after that.
